sponsorship_compassion/model/contracts.py

```diff
@@ -957,115 +352,119 @@ def suspend_contract(self, cr, uid, ids, context=None, date_start=None,
-            contract.child_id.write(child_vals)
-
-        # Delete workflow for this contract
+    def _on_contract_active(self, cr, uid, ids, context=None):
+        """ Hook for doing something when contract is activated.
+        Update child to mark it has been sponsored, and update partner
+        to add the 'Sponsor' category.
+        """
+        # Read data in english
+        if context is None:
+            context = {}
+        ctx = context.copy()
+        ctx['lang'] = 'en_US'
         wf_service = netsvc.LocalService('workflow')
-        wf_service.trg_delete(uid, 'recurring.contract', contract_id, cr)
-        logger.info("Contract " + str(contract_id) + " terminated.")
-        return True
-
-
-# just to modify access rights...
-class recurring_invoicer(orm.Model):
-    _inherit = 'recurring.invoicer'
-    _name = 'recurring.invoicer'
-
-
-class account_invoice(orm.Model):
-    _inherit = 'account.invoice'
-    _name = 'account.invoice'
-
-
-class contract_line(orm.Model):
-    _inherit = 'recurring.contract.line'
-
-
-class account_move(orm.Model):
-    _inherit = 'account.move'
-
-
-class account_move_line(orm.Model):
-    _inherit = 'account.move.line'
-
-
-class payment_line(orm.Model):
-    _inherit = 'payment.line'
-
-
-class account_journal(orm.Model):
-    _inherit = 'account.journal'
+        if not isinstance(ids, list):
+            ids = [ids]
+        sponsor_cat_id = self.pool.get('res.partner.category').search(
+            cr, uid, [('name', '=', 'Sponsor')], context=ctx)[0]
+        for contract in self.browse(cr, uid, ids, context=ctx):
+            if contract.child_id:
+                contract.child_id.write({'has_been_sponsored': True})
+                partner_categories = set(
+                    [cat.id for cat in contract.partner_id.category_id
+                     if cat.name != 'Old Sponsor'])
+                partner_categories.add(sponsor_cat_id)
+                # Standard way in Odoo to set one2many fields
+                contract.partner_id.write({
+                    'category_id': [(6, 0, list(partner_categories))]})
+            wf_service.trg_validate(
+                uid, self._name, contract.id,
+                'contract_active', cr)
+            logger.info("Contract " + str(contract.id) + " activated.")
 
+    def _on_change_child_id(self, cr, uid, ids, child_id, partner_id=None,
+                            context=None):
+        """Link/unlink child to sponsor
+        """
+        for contract in self.browse(cr, uid, ids, context):
+            if contract.child_id and contract.child_id != child_id:
+                # Free the previously selected child
+                contract.child_id.write({'sponsor_id': False})
+            if child_id:
+                # Mark the selected child as sponsored
+                self.pool.get('compassion.child').write(
+                    cr, uid, child_id, {
+                        'sponsor_id': partner_id or contract.partner_id.id},

```

Comment:
We will rather take the correspondant. So replace by vals.get('correspondant_id') or contract.correspondant_id.id.

You can then replace child_id and partner_id from method parameters by vals (all vals updated).